### PR TITLE
[fix] Fix issue #25 #26 #27

### PR DESCRIPTION
**Pull request by magikcypress**

### CODE REVIEW

**Comment by magikcypress** [comment]
This PR needs to be tested

**Comment by JimboJoe** [comment]
... and you need to change de `check_process` file accordingly (is_public as a boolean).

**Comment by JimboJoe** [comment]
Weirdly, package_check still fails here (Error: yunohost cli - Invalid choice for argument 'is_public', it must be one of true, false)....

**Comment by magikcypress** [comment]
@JimboJoe Thanks ;) the choice should not be made in the manifest.json

fix 46a68e2

File: scripts/install [hunk 1]
```diff
@@ -83,19 +87,16 @@ fi
...
+fi
 
-if [ "$is_public" = "No" ];
+if [ $is_public -eq 1 ]
 then
-	# Exit public access
-	ynh_app_setting_delete $app unprotected_uris
-	sudo yunohost app ssowatconf
+	ynh_app_setting_set $app skipped_uris "/" # Make app public if necessary
 fi
```
Comment (JimboJoe) [inline scripts/install]: Why changing this existing (tested) logic based on unprotected_uris to use skipped_uris (about which I'm very doubtful)?

File: scripts/restore [hunk 2]
```diff
@@ -41,6 +48,9 @@ if [ -f $phpfpm_ini ]; then
 	ynh_die "The PHP FPM INI configuration already exists at '${phpfpm_ini}'. You should safely delete it before restoring this app."
 fi
 
+# Create system user dedicace for this app
+ynh_system_user_create $app
+
  # Restore sources & data
 sudo cp -a ./sources "${final_path}"
```
Comment (JimboJoe) [inline scripts/restore]: chown should change owner to $app

File: scripts/upgrade [hunk 3]
```diff
@@ -82,17 +91,18 @@ done
...
+	sudo chown -R $app:root $final_path/{data/attic,data/cache,data/index,data/locks,data/media*,data/meta,data/pages,data/tmp}
 fi
-sudo chown -R www-data:root $final_path/{conf,data,lib/plugins,lib/tpl}
+sudo chown -R $app:root $final_path/{conf,data,lib/plugins,lib/tpl}
 sudo chmod -R 700 $final_path/conf
 sudo chmod -R 700 $final_path/data
 sudo chmod -R 700 $final_path/lib/plugins
-sudo chmod -R 700 $final_path/lib/tpl
+# sudo chmod -R 700 $final_path/lib/tpl
+sudo chmod 775 $final_path/lib/tpl/{dokuwiki,dokuwiki/images}
```
Comment (JimboJoe) [inline scripts/upgrade]: Same remark and question here (as in install script)

File: scripts/upgrade [hunk 4]
```diff
@@ -107,13 +117,14 @@ fi
...
+# Set ssowat config
+if [ "$is_public" = "Yes" ]; then
+	ynh_app_setting_set $app is_public 1 # Fixe is_public en booléen
+	is_public=1
+elif [ "$is_public" = "No" ]; then
+	ynh_app_setting_set $app is_public 0
+ 	is_public=0
 fi
 
-sudo systemctl reload php5-fpm
```
Comment (JimboJoe) [inline scripts/upgrade]: I don't see why this wouldn't be needed any more... Some user may still use the previous version which wasn't using a dedicated php5-fpm.

File: scripts/upgrade [hunk 5]
```diff
@@ -107,13 +117,14 @@ fi
...
-# Setup SSOwat
-ynh_app_setting_set "$app" is_public "$is_public"
-if [ "$is_public" = "Yes" ];
-then
-  ynh_app_setting_set "$app" unprotected_uris "/"
+# Set ssowat config
+if [ "$is_public" = "Yes" ]; then
+	ynh_app_setting_set $app is_public 1 # Fixe is_public en booléen
+	is_public=1
+elif [ "$is_public" = "No" ]; then
```
Comment (JimboJoe) [inline scripts/upgrade]: you could simplify by using an `else` instead of `elif` (binary case).

File: scripts/install [hunk 6]
```diff
@@ -57,17 +60,18 @@ SETUP_SOURCE
...
 
 # except for conf, data, some data subfolders, and lib/plugin, where www-data must have write permissions
-sudo chown -R www-data:root $final_path/{conf,data,data/attic,data/cache,data/index,data/locks,data/media*,data/meta,data/pages,data/tmp,lib/plugins,lib/tpl}
+sudo chown -R $app:root $final_path/{conf,data,data/attic,data/cache,data/index,data/locks,data/media*,data/meta,data/pages,data/tmp,lib/plugins,lib/tpl}
 sudo chmod -R 700 $final_path/conf
 sudo chmod -R 700 $final_path/data
 sudo chmod -R 700 $final_path/lib/plugins
-sudo chmod -R 700 $final_path/lib/tpl
+# sudo chmod -R 700 $final_path/lib/tpl
+sudo chmod 775 $final_path/lib/tpl/{dokuwiki,dokuwiki/images}
```
Comment (magikcypress) [inline scripts/install]: Yes, because a pictures is not display on interface.

File: scripts/upgrade (hunk 4 above)
Comment (magikcypress) [inline scripts/upgrade]: php5-fpm is reload into POOL_FPM function ;)

File: scripts/install (hunk 1 above)
Comment (magikcypress) [inline scripts/install]: I use unprotected_uris

File: scripts/upgrade (hunk 5 above)
Comment (magikcypress) [inline scripts/upgrade]: fixed. Thanks ;)

**Comment by maniackcrudelis** [comment]
I've tested this PR, still some errors.

- userdel: user dokuwiki is currently used by process xxx
I think you should reload php-fpm, which use this user, before removing it

- I've made a test for the upgrade from the current version.
And, the script work fine. But, I have a wonderful 502 :D

Except that, package check is clean.

**Comment by magikcypress** [comment]
Thank Maniack. 
i fixed error 502, the socket for php5-fpm was not correct in the nginx configuration.

**Comment by maniackcrudelis** [comment]
All clean on package check.
And the upgrade works fine now.

**Comment by JimboJoe** [comment]
I upgraded my production instance and there are still problems: images from plugins (move,  indexmenu for example) aren't served (403 forbidden). I had to `chmod -R go+rx lib/plugins/` in order to make it work. I'm sure it's not the right solution, though...
I'm still a bit puzzled by this separate php-fpm user; still, files served by nginx must be reachable by the nginx process, so by www-data...?

**Comment by maniackcrudelis** [comment]
The matter about the user must be only a matter of write access.
The good manner is to keep all your files under root, and give write access only on the files who really need it.
Mostly plugin I think, for example.

For that, I'm wondering why you use that kind of access.
https://github.com/magikcypress/dokuwiki_ynh/blob/master/scripts/install#L67-L71
It's better, and more simple to keep all your files under 755, eventually 700 for a file who contains some sensitive data.
And set all of that to `root:`, and `$app:root` for files who need write access or restricted access.

Nginx need only a read access. Never any write access. It's why you can use a dedicated user with php-fpm

**Comment by JimboJoe** [comment]
I guess my problem with access rights to plugins images will persist because of this: https://github.com/YunoHost-Apps/dokuwiki_ynh/pull/28/commits/ff48bfda3525211a3ccf5e876690b331a0b3aceb#diff-44cb16c778719320333118c04d509a7cL101
nginx runs as www-data, so it still lacks read right on lib/plugins.

**Comment by maniackcrudelis** [comment]
Did you try it?
Maybe it lacks a -R but with 755, even if nginx run under www-data, it can reads it.

**Comment by JimboJoe** [comment]
No I didn't test it, but I was referring to line 101 and this: `sudo chmod -R 700 $final_path/lib/plugins`  so for me it still lacks some read rights :-)

**Comment by magikcypress** [comment]
I added 755 for the plugin directory.

**Comment by alexAubin** [comment]
(Bump ? Sounds like this is needed for https://github.com/YunoHost/apps/pull/209)

**Comment by magikcypress** [comment]
Yep!  YunoHost/apps#209 need to approved

**Pull request merged**